### DISCUSSION TITLE
magit-log-mode-hook Does Not Work

### DISCUSSION THREAD

**Discussion by gthomsen** [magit/magit#5072] [Support] [answered]
I'm having difficulty adding a function to the `magit-log-mode-hook` hook with Magit 3.3.0, Git 2.34.1, and Emacs 27.1 on gnu/linux.  With the following code, it appears that `turn-off-whitespace-mdoe` is never called from the `magit-log-mode-hook`:

```lisp
(defun turn-off-whitespace-mode ()
  "Disable whitespace-mode in the current buffer."
  (interactive)
  (whitespace-mode -1))

(eval-after-load "magit"
  '(progn
     (magit-add-section-hook 'magit-status-sections-hook 'turn-off-whitespace-mode)
     (magit-add-section-hook 'magit-revision-sections-hook 'turn-off-whitespace-mode)

     ;; NOTE: this doesn't seem to work with add-hook or magit-add-section-hook.
     (add-hook 'magit-log-mode-hook 'turn-off-whitespace-mode)))
```

The following `init.el` leaves `whitespace-mode` enabled when showing the commit history of a repository (`M-x magit RET` and then `l l`) with an empty Emacs configuration and a fresh installation of Magit 3.3.0 from Melpa Stable:

```lisp
(require 'package)
(add-to-list 'package-archives
  ("melpa-stable" . "http://stable.melpa.org/packages/") t)

(require 'whitespace)

(global-whitespace-mode t)

(defun turn-off-whitespace-mode ()
  "Disable whitespace-mode in the current buffer."
  (interactive)
  (whitespace-mode -1))

(eval-after-load "magit"
  '(progn
     (magit-add-section-hook 'magit-status-sections-hook 'turn-off-whitespace-mode)
     (magit-add-section-hook 'magit-revision-sections-hook 'turn-off-whitespace-mode)
     (add-hook 'magit-log-mode-hook 'turn-off-whitespace-mode)))
```

The fact that the hook did not run can be seen by running `M-x turn-off-whitespace-mode`.

Please advise if `magit-log-mode-hook` is the wrong hook to use for commit history.

**Comment by tarsius** [comment, marked as answer]
> ```elisp
> (magit-add-section-hook 'magit-status-sections-hook 'turn-off-whitespace-mode)
> ```

Don't do that. `magit-add-section-hook` is only for adding functions that *insert sections* to hook variables intended for holding a list of such functions. `turn-off-whitespace-mode` doesn't insert sections, so (1) using `magit-add-section-hook` is wrong and (2) adding it to `magit-section-sections-hook` is wrong.

> The fact that the hook did not run can be seen by running `M-x turn-off-whitespace-mode`.

The fact that `magit-log-mode-hook` is run when `magit-log-mode` is turned on can be observed with this:

```elisp
(defun hello-hook-fn () (message "Hello %s" major-mode))

(add-hook 'magit-log-mode-hook 'hello-hook-fn)
```

The problem is that when `global-whitespace-mode` is enabled, it does its thing after the hooks on any major-mode's hook. So the local gets disabled first by your hook function (it actually is disabled at this point, so your hook doesn't actually have an effect). Then `global-whitespace-mode` checks whether, based on its own configuration, it should enable the local mode, and unfortunately with its default configuration, the answer is yes, in nearly every buffer/mode, including magit buffers/modes.

To prevent that you have to either refrain from using `global-whitespace-mode` (and enable the local mode directly, where ever you actually want that) or configure the global mode according to your preferences. See `whitespace-global-modes`, `whitespace-enable-predicate` and `whitespace-turn-on-if-enabled`.

**Comment by gthomsen** [comment]
Thanks for a quick response and the details on both how the Magit hooks work, as well as the key parts of `whitespace-mode` that I needed to understand.  This configuration seems to do what I want while leaving `global-whitespace-mode` on:

```lisp
(setq whitespace-global-modes '(not magit-log-mode
                                    magit-revision-mode
                                    magit-status-mode))
```
